cmd/pulumi.go

```diff
@@ -155,6 +175,170 @@ func NewPulumiCmd() *cobra.Command {
 	return cmd
 }
 
+// checkForUpdate checks to see if the CLI needs to be updated, and if so emits a warning, as well as information
+// as to how it can be upgraded.
+func checkForUpdate() {
+	curVer, err := semver.ParseTolerant(version.Version)
+	if err != nil {
+		glog.V(3).Infof("error parsing current version: %s", err)
+	}
+
+	// We don't care about warning for you to update if you have installed a developer version
+	if isDevVersion(curVer) {
+		return
+	}
+
+	latestVer, oldestAllowedVer, err := getCLIVersionInfo()
+	if err != nil {
+		glog.V(3).Infof("error fetching latest version information: %s", err)
+	}
+
+	if oldestAllowedVer.GT(curVer) {
+		cmdutil.Diag().Warningf(diag.RawMessage("", getUpgradeMessage(latestVer, curVer)))
+	}
+}
+
+// getCLIVersionInfo returns information about the latest version of the CLI and the oldest version that should be
+// allowed without warning. It caches data from the server for a day.
+func getCLIVersionInfo() (semver.Version, semver.Version, error) {
+	latest, oldest, err := getCachedVersionInfo()
+	if err == nil {
+		return latest, oldest, err
+	}
+
+	client := client.NewClient(cloud.DefaultURL(), "")
+	latest, oldest, err = client.GetCLIVersionInfo(commandContext())
+	if err != nil {
+		return semver.Version{}, semver.Version{}, err
+	}
+
+	err = cacheVersionInfo(latest, oldest)
+	if err != nil {
+		glog.V(3).Infof("failed to cache version info: %s", err)
+	}
+
+	return latest, oldest, err
+}
+
+// cacheVersionInfo saves version information in a cache file to be looked up later.
+func cacheVersionInfo(latest semver.Version, oldest semver.Version) error {
+	updateCheckFile, err := workspace.GetCachedVersionFilePath()
+	if err != nil {
+		return err
+	}
+
+	file, err := os.OpenFile(updateCheckFile, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0600)
+	if err != nil {
+		return err
+	}
+	defer contract.IgnoreClose(file)
+
+	return json.NewEncoder(file).Encode(cachedVersionInfo{
+		LatestVersion:        latest.String(),
+		OldestWithoutWarning: oldest.String(),
+	})
+}
+
+// getCachedVersionInfo reads cached information about the newest CLI version, returning the newest version avaliaible
+// as well as the oldest version that should be allowed without warning the user they should upgrade.
```

Comment:
Nit: Consider adding a description of what would cause an error, since reading the code I wouldn't have expected it to fail if the file was too old. Perhaps:

"Returns any IO errors, or if the cache file's modification is too old (i.e. cache is stale)."